### ISSUE TITLE
Setting globalText property on SuggestBuilder instead of the #text property on each SuggestionBuilder returns null from SearchResponse#getSuggest

### ISSUE THREAD

**Issue by etay2000** [elastic/elasticsearch#23340]
**Elasticsearch version**: 5.2.1

**Description of the problem including expected versus actual behavior**:

Calling the setGlobalText method on SuggestBuilder should be a shortcut to setting the text() property on each SuggestionBuilder added to the SuggestBuilder. Currently setting just the global text causes the Suggest object returned by SearchResponse to be null.

**Steps to reproduce**:
 `
String searchText = "foo";
SuggestBuilder suggestBuilder = new SuggestBuilder(); 
suggestBuilder.setGlobalText(searchText);

SuggestionBuilder<?> suggestionBuilder = SuggestBuilders.completionSuggestion("fieldName");
// suggestionBuilder.text(searchText); 
// uncomment the line above and the searchResponse#getResponse will exist as expected
suggestBuilder.addSuggestion("suggestion-name", suggestionBuilder);

searchResponse.getSuggest() // This is null when only the global text is set.
`
The Suggest instance should not be null if the global text is set on the SuggestBuilder.

**Comment by etay2000** [comment]
Directly above the section in the docs you referenced, this is mentioned:

> To avoid repetition of the suggest text, it is possible to define a global text. In the example below the suggest text is defined globally and applies to the my-suggest-1 and my-suggest-2 suggestions.

```
 POST _search
{
  "suggest": {
    "text" : "tring out Elasticsearch",
    "my-suggest-1" : {
      "term" : {
        "field" : "message"
      }
    },
    "my-suggest-2" : {
       "term" : {
        "field" : "user"
       }
    }
  }
}
```
This leads me to believe that setting the global text is a shortcut for setting the text() property on each SuggestionBuilder. 

Also, in the code you referenced:
```
if (suggestionContext.getText() == null) {
    if (globalText == null) {
        throw new IllegalArgumentException("The required text option is missing");
    }
    suggestionContext.setText(BytesRefs.toBytesRef(globalText));
}
```
I read that as: 
If the suggestionContext does not have it's text() property set, and globalText is not null, then use the globalText to assign the the text() property of the suggestionContext.

Otherwise, what would be the point of setting the globalText if you still had to set the text property of each SuggestionBuilder?

**Comment by cbuescher** [comment]
@etay2000 The docs refer to the Rest syntax, not the Java API. Your reading of the code is correct, but here we are already taking about internals (SuggestionContext, SuggestionSearchContext), not the builders. The `globalText` will be used when set on the SuggestBuilder and no `text` from the individual Suggestion overwrite it. But that doesn't mean we need to transfer this to the SuggestionBuilders.

**Comment by etay2000** [comment]
@cbuescher I assumed the Rest syntax would mirror the Java API (or vice versa) with regards to setting the globalText.  I am not too familiar with the internals, so I am still a bit confused by when and why the globalText would actually be used. My interpretation of the benefit of the globalText option was to prevent the repetition of setting the same suggest text. As it stands in the Java API, each SuggestionBuilder is still required to have its text property set even if it shares the same text value with all of the other SuggestionBuilders. If the text property is not set, then why wouldn't the globalText propagate down to it? This to me seems to defeat the purpose of having an option of setting the globalText on the SuggestBuilder in the first place.

**Comment by cbuescher** [comment]
> As it stands in the Java API, each SuggestionBuilder is still required to have its text property set even if it shares the same text value with all of the other SuggestionBuilders. If the text property is not set, then why wouldn't the globalText propagate down to it?

Have you tried setting the `globalText` on the SuggestBuilder and then using this for the seach? It should work regardless. You don't have to set the `text` on each SuggestionBuilder, these things are merged later (the place I pointed to)...

**Comment by etay2000** [comment]
Yes that was my point in my first post (sorry for the poor code formatting). If I set the globalText on the SuggestBuilder but not on any SuggestionBuilders I get a null Suggest instance back from the SearchResponse. However, if I set the suggest text on the SuggestionBuilders I get a proper Suggest instance as expected.

```String searchText = "foo";
SuggestBuilder suggestBuilder = new SuggestBuilder();
suggestBuilder.setGlobalText(searchText);

SuggestionBuilder<?> suggestionBuilder = SuggestBuilders.completionSuggestion("fieldName");
// suggestionBuilder.text(searchText);
// As is, searchResponse#getSuggest will return null,
// uncomment the line above and the searchResponse#getSuggest will return as expected
suggestBuilder.addSuggestion("suggestion-name", suggestionBuilder);```

**Comment by cbuescher** [comment]
@etay2000 thanks for the clarification, I was able to reproduce this now in a simple test. The problem seems to be limited to completion suggestions where neither `prefix`, `regex` nor `text`  specified on the individual suggestion. In this case we internally overwrite the `text` property with the global text set on the Suggest builder. The problem is that internally the completion suggester need either `prefix` or `regex` to be set and throws an error. There should be shard failures in the response you are receiving. 
I think in case of missing `text` on the suggestion, we should use the global text to also overwrite the prefix used by the CompletionSuggester. I will open a PR shortly.

**Comment by etay2000** [comment]
@cbuescher Sure enough there were shard failures and somehow I missed the IllegalArgumentException("'prefix' or 'regex' must be defined") on the stack, so your fix should definitely take care of it. Thanks for the quick response.

**Comment by cbuescher** [comment]
Closed by #23451
